torch/distributed/_shard/partial_tensor.py

```diff
@@ -127,25 +126,49 @@ def reshard(self, resharding_spec: shard_spec.ShardingSpec) -> ShardedTensor:
             raise NotImplementedError("Only real partial tensor supported for reshard.")
         sharding_dim = int(resharding_spec.dim)  # type: ignore[attr-defined]
         chunk_mode_res = self.local_shard.size(sharding_dim) % self.process_group.size()
+        local_shard = self.local_shard
         # Add padding when the size is not divisible by the world size.
         if chunk_mode_res != 0:
-            padding = [0] * (len(self.local_shard.size()) * 2)
+            padding = [0] * (self.local_shard.dim() * 2)
             padding[-1] = self.process_group.size() - chunk_mode_res
-            self.local_shard = torch.nn.functional.pad(
+            local_shard = torch.nn.functional.pad(
                 self.local_shard,
                 tuple(padding),
                 "constant",
                 0,
             )
-
-        local_shards = self.local_shard.chunk(
-            self.process_group.size(), dim=sharding_dim
-        )
+        current_rank = dist.get_rank(self.process_group)  # type: ignore[attr-defined]
+        rank_idx = None
+        rearrange_local_shards = False
+        indices = [0] * self.process_group.size()
+        for idx, placement in enumerate(resharding_spec.placements):  # type: ignore[attr-defined]
+            if placement.rank() == current_rank:  # type: ignore[index]
```

Comment:
Is this needed because the existing code didn't work with rearrangements? If so, is there a unit test covering this case now?

Reply:
Previously we didn't consider that.. And yes, I added new unit test cases for it in test_partial_tensor.